File: stack/394.py
class Solution:
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        stack = []
        multi = 0
        ret = ''
        for c in s:
            if c == '[':
                stack.append(ret)
                stack.append(multi)
                ret = ''
                multi = 0
            elif c == ']':
                num = stack.pop()
                temp = stack.pop()
                ret = temp + num*ret
            elif c.isdigit():
                multi = multi*10 + int(c)
            else:
                ret += c
        return ret        
        

File: stack/394.cc
class Solution {
public:
    string decodeString(string s) {
        std::stack<std::string> s_;
        int index = 0;
        while (index < s.size()) {
            if (s[index] == ']') {
                std::string temp = "";
                while (s_.top() != "[") {
                    temp = s_.top() + temp;
                    s_.pop();
                }
                // find "["
                s_.pop();
                std::string repeat = "";
                while (!s_.empty() && s_.top() >= "0" && s_.top() <= "9") {
                    repeat = s_.top() + repeat;
                    s_.pop();
                }
                int repeat_num = std::stoi(repeat);
                repeat = "";
                while (repeat_num > 0) {
                    repeat += temp;
                    --repeat_num;
                }
                s_.emplace(repeat);
            } else {
                s_.emplace(string(1,s[index]));
            }
            ++index;
        }
        std::string ret = "";
        while (!s_.empty()) {
            ret = s_.top() + ret;
            s_.pop();
        }
        return ret;
    }
};
